test/unit/data-connect/data-connect-api-client-internal.spec.ts

```diff
@@ -218,10 +224,382 @@ describe('DataConnectApiClient', () => {
+        .to.be.rejectedWith(FirebaseDataConnectError, /`data` must be a non-null object./);
+    });
+
+    it('should throw FirebaseDataConnectError for array data', async () => {
+      await expect(apiClient.upsert(tableName, [{ data: 1 }]))
+        .to.be.rejectedWith(FirebaseDataConnectError, /`data` must be an object, not an array, for single upsert./);
+    });
+
+    it('should amend the message for query errors', async () => {
+      await expect(apiClientQueryError.upsert(tableName, { data: 1 }))
+        .to.be.rejectedWith(FirebaseDataConnectError, `${serverErrorString}. ${additionalErrorMessageForBulkImport}`);
+    });
+  });
+
+  // --- UPSERT MANY TESTS ---
+  describe('upsertMany()', () => {
+    tableNames.forEach((tableName, index) => {
+      const expectedMutation = `mutation { ${formatedTableNames[index]}_upsertMany(data: [{ name: "a" }]) }`;
+      it(`should use the formatted tableName in the gql query: "${tableName}" as "${formatedTableNames[index]}"`,
+        async () => {
+          await apiClient.upsertMany(tableName, [{ name: 'a' }]);
+          await expect(executeGraphqlStub).to.have.been.calledOnceWithExactly(normalizeGraphQLString(expectedMutation));
+        });
+    });
+
+    it('should call executeGraphql with the correct mutation for simple data array', async () => {
+      const simpleDataArray = [{ id: 'k1' }, { id: 'k2', value: 99 }];
+      const expectedMutation = `
+      mutation { ${formatedTableName}_upsertMany(data: [{ id: "k1" }, { id: "k2", value: 99 }]) }`;
+      await apiClient.upsertMany(tableName, simpleDataArray);
+      expect(executeGraphqlStub).to.have.been.calledOnceWithExactly(normalizeGraphQLString(expectedMutation));
+    });
+
+    it('should call executeGraphql with the correct mutation for complex data array', async () => {
+      const complexDataArray = [
+        { id: 'x', active: true, info: { nested: 'n1/\\"x' } },
+        { id: 'y', scores: [null, 2] }
+      ];
+      const expectedMutation = `
+      mutation { ${formatedTableName}_upsertMany(data: 
+      [{ id: "x", active: true, info: { nested: "n1/\\\\\\"x" } }, { id: "y", scores: [null, 2] }]) }`;
+      await apiClient.upsertMany(tableName, complexDataArray);
+      expect(executeGraphqlStub).to.have.been.calledOnceWithExactly(normalizeGraphQLString(expectedMutation));
+    });
+
+    it('should call executeGraphql with the correct mutation for undefined and null', async () => {
+      const dataArray = [
+        dataWithUndefined,
+        dataWithUndefined
+      ]
+      const expectedMutation = `
+      mutation {
+      ${formatedTableName}_upsertMany(data: [{
+       genre: "Action",
+       title: "Die Hard",
+       ratings: null,
+       director: {},
+       extras: [1, null, "hello", null, { a: 1 }]
+      },
+      {
+       genre: "Action",
+       title: "Die Hard",
+       ratings: null,
+       director: {},
+       extras: [1, null, "hello", null, { a: 1 }]
```

Comment:
We shouldn't allow for sparsed arrays, can we catch this somehow?

Reply:
Good catch! Made some changes to catch this and throw before we make the request to the server. Added new test cases to cover this change. Thanks!

Reply:
Discussed offline and decided not to validate this on the SDK side. We will let the BE service handle this for now.